src/cmds/restic/cmd_forget.go

```diff
@@ -114,38 +115,36 @@ func runForget(opts ForgetOptions, gopts GlobalOptions, args []string) error {
 		return err
 	}
 
-	// parse arguments as hex strings
-	var ids []string
-	for _, s := range args {
-		_, err := hex.DecodeString(s)
-		if err != nil {
-			Warnf("argument %q is not a snapshot ID, ignoring\n", s)
-			continue
-		}
-
-		ids = append(ids, s)
+	// group by hostname and dirs
+	type key struct {
+		Hostname string
+		Dirs     string
 	}
+	snapshotGroups := make(map[key]restic.Snapshots)
 
-	// process all snapshot IDs given as arguments
-	for _, s := range ids {
-		id, err := restic.FindSnapshot(repo, s)
-		if err != nil {
-			Warnf("could not find a snapshot for ID %q, ignoring\n", s)
-			continue
-		}
-
-		if !opts.DryRun {
-			h := restic.Handle{Type: restic.SnapshotFile, Name: id.String()}
-			err = repo.Backend().Remove(h)
-			if err != nil {
-				return err
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args) {
+		if len(args) > 0 {
+			// When explicit snapshots args are given, remove them immediately.
+			if !opts.DryRun {
+				h := restic.Handle{Type: restic.SnapshotFile, Name: sn.ID().String()}
+				if err = repo.Backend().Remove(h); err != nil {
+					return err
+				}
+				Verbosef("removed snapshot %v\n", sn.ID())
+			} else {
+				Verbosef("would have removed snapshot %v\n", sn.ID())
 			}
-
-			Verbosef("removed snapshot %v\n", id.Str())
 		} else {
-			Verbosef("would remove snapshot %v\n", id.Str())
+			// Not cool, we lose the benefit of yielding, e.g. only one snapshot in memory.
```

Comment:
That using a channel to hand over a Snapshot one by one is completely defeated by putting them all in a list. But if you don't like the remainder, I can certainly make it go the way of the dodo's ^^